File: be/src/main/java/com/example/mingle/domain/post/legalpost/service/ContractService.java
package com.example.mingle.domain.post.legalpost.service;

import com.example.mingle.domain.admin.panel.dto.ContractConditionResponse;
import com.example.mingle.domain.admin.panel.dto.ContractResponse;
import com.example.mingle.domain.admin.panel.dto.ContractSearchCondition;
import com.example.mingle.domain.admin.panel.service.ContractSpecification;
import com.example.mingle.domain.post.legalpost.dto.contract.CreateContractRequest;
import com.example.mingle.domain.post.legalpost.dto.contract.CreateInternalContractRequest;
import com.example.mingle.domain.post.legalpost.dto.contract.UpdateContractRequest;
import com.example.mingle.domain.post.legalpost.entity.Contract;
import com.example.mingle.domain.post.legalpost.entity.InternalContract;
import com.example.mingle.domain.post.legalpost.entity.SettlementRatio;
import com.example.mingle.domain.post.legalpost.enums.*;
import com.example.mingle.domain.post.legalpost.repository.ContractRepository;
import com.example.mingle.domain.post.legalpost.repository.InternalContractRepository;
import com.example.mingle.domain.post.legalpost.repository.SettlementRatioRepository;
import com.example.mingle.domain.user.team.entity.ArtistTeam;
import com.example.mingle.domain.user.team.repository.ArtistTeamRepository;
import com.example.mingle.domain.user.user.entity.User;
import com.example.mingle.domain.user.user.repository.UserRepository;
import com.example.mingle.global.aws.AwsS3Uploader;

import com.example.mingle.global.exception.ApiException;
import com.example.mingle.global.exception.ErrorCode;
import com.example.mingle.global.security.auth.SecurityUser;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.math.BigDecimal;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class ContractService {

    private final ContractRepository contractRepository;
    private final SettlementRatioRepository ratioRepository;
    private final UserRepository userRepository;
    private final ArtistTeamRepository teamRepository;
    private final AwsS3Uploader s3Uploader;
    private final DocusignService docusignService;
    private final InternalContractRepository internalContractRepository;

    public Long createContract(CreateContractRequest req, MultipartFile file) throws IOException {
        User user1 = userRepository.findById(req.userId()).orElseThrow();
        ArtistTeam team = teamRepository.findById(req.teamId()).orElse(null);
        String fileUrl = s3Uploader.upload(file, "contracts");

        Contract contract = new Contract();
        contract.setUser(user1);
        contract.setTeam(team);
        contract.setFileUrl(fileUrl);
        contract.setSummary(req.summary());
        contract.setTitle(req.title());
        contract.setCompanyName("Mingle");
        contract.setContractCategory(req.contractCategory());
        contract.setStartDate(req.startDate());
        contract.setEndDate(req.endDate());
        contract.setStatus(ContractStatus.DRAFT);
        contract.setContractType(req.contractType());
        contract.setContractAmount(req.contractAmount());

        contractRepository.save(contract);

        if (req.useManualRatios()) {
            //  수동 입력 방식
            for (CreateContractRequest.SettlementRatioDto dto : req.ratios()) {
                SettlementRatio ratio = new SettlementRatio();

                if (dto.userId() != null) {
                    User user = userRepository.findById(dto.userId()).orElseThrow();
                    ratio.setUser(user);
                    contract.getParticipants().add(user);
                } else {
                    ratio.setUser(null); // 회사 몫
                }
                ratio.setContract(contract);
                ratio.setRatioType(dto.ratioType());
                ratio.setPercentage(dto.percentage());
                ratioRepository.save(ratio);
            }
            contractRepository.save(contract);

        } else {
            // 내부계약 기반 자동 방식
            BigDecimal sum = BigDecimal.ZERO;

            // 1. 유저 기반 비율 저장 (아티스트/프로듀서)
            for (Long userId : req.targetUserIds()) {
                User user = userRepository.findById(userId).orElseThrow();
                contract.getParticipants().add(user);

                InternalContract internal = internalContractRepository
                        .findValidByUserAndDate(user, LocalDate.now(), ContractStatus.TERMINATED)
                        .orElseThrow(() -> new IllegalStateException("내부 계약 없음"));

                BigDecimal userRatio = internal.getDefaultRatio();
                sum = sum.add(userRatio);

                SettlementRatio ratio = new SettlementRatio();
                ratio.setContract(contract);
                ratio.setRatioType(internal.getRatioType()); // ARTIST, PRODUCER 등
                ratio.setUser(user);
                ratio.setPercentage(userRatio);
                ratioRepository.save(ratio);
            }

            // 2. 회사 몫 자동 계산 (100 - 참여자 비율 합계)
            BigDecimal companyRatio = BigDecimal.valueOf(100).subtract(sum);
            if (companyRatio.compareTo(BigDecimal.ZERO) > 0) {
                SettlementRatio agencyRatio = new SettlementRatio();
                agencyRatio.setContract(contract);
                agencyRatio.setRatioType(RatioType.AGENCY); // 회사 몫
                agencyRatio.setUser(null); // 유저 없음
                agencyRatio.setPercentage(companyRatio);
                ratioRepository.save(agencyRatio);
            }
            contractRepository.save(contract);
        }

        return contract.getId();
    }

    // 내부 계약 생성
    public Long create(CreateInternalContractRequest req, MultipartFile file) throws IOException {
        User user = userRepository.findById(req.userId())
                .orElseThrow(() -> new EntityNotFoundException("사용자 없음"));

        String fileUrl = s3Uploader.upload(file, "internal-contracts"); // S3 경로 등

        InternalContract contract = new InternalContract();
        contract.setUser(user);
        contract.setRatioType(req.ratioType());
        contract.setDefaultRatio(req.defaultRatio());
        contract.setStartDate(req.startDate());
        contract.setEndDate(req.endDate());
        contract.setStatus(ContractStatus.DRAFT);
        contract.setFileUrl(fileUrl);

        internalContractRepository.save(contract);
        return contract.getId();
    }

    @Transactional
    public Long updateContract(Long contractId, UpdateContractRequest req, MultipartFile file) throws IOException {
        Contract contract = contractRepository.findById(contractId)
                .orElseThrow(() -> new EntityNotFoundException("계약서 없음"));

        // 파일이 있을 경우에만 업데이트
        if (file != null && !file.isEmpty()) {
            String fileUrl = s3Uploader.upload(file, "contracts");
            contract.setFileUrl(fileUrl);
        }

        // 기본 필드 수정
        contract.setTitle(req.title());
        contract.setSummary(req.summary());
        contract.setStartDate(req.startDate());
        contract.setEndDate(req.endDate());
        contract.setContractAmount(req.contractAmount());
        contract.setContractCategory(req.contractCategory());
        contract.setContractType(req.contractType());
        contract.setStatus(req.status());

        // 기존 SettlementRatio/참여자 초기화
        ratioRepository.deleteAllByContract(contract);
        contract.getParticipants().clear();

        if (req.useManualRatios()) {
            // 수동 입력 방식
            for (UpdateContractRequest.SettlementRatioDto dto : req.ratios()) {
                SettlementRatio ratio = new SettlementRatio();

                if (dto.userId() != null) {
                    User user = userRepository.findById(dto.userId()).orElseThrow();
                    ratio.setUser(user);
                    contract.getParticipants().add(user);
                } else {
                    ratio.setUser(null); // 회사 몫
                }
                ratio.setContract(contract);
                ratio.setRatioType(dto.ratioType());
                ratio.setPercentage(dto.percentage());

                ratioRepository.save(ratio);
            }

        } else {
            // 내부 계약 기반 자동 방식
            BigDecimal sum = BigDecimal.ZERO;

            for (Long userId : req.targetUserIds()) {
                User user = userRepository.findById(userId).orElseThrow();
                contract.getParticipants().add(user);

                InternalContract internal = internalContractRepository
                        .findValidByUserAndDate(user, LocalDate.now(), ContractStatus.TERMINATED)
                        .orElseThrow(() -> new IllegalStateException("내부 계약 없음"));

                BigDecimal userRatio = internal.getDefaultRatio();
                sum = sum.add(userRatio);

                SettlementRatio ratio = new SettlementRatio();
                ratio.setContract(contract);
                ratio.setRatioType(internal.getRatioType());
                ratio.setUser(user);
                ratio.setPercentage(userRatio);

                ratioRepository.save(ratio);
            }

            BigDecimal companyRatio = BigDecimal.valueOf(100).subtract(sum);
            if (companyRatio.compareTo(BigDecimal.ZERO) > 0) {
                SettlementRatio agencyRatio = new SettlementRatio();
                agencyRatio.setContract(contract);
                agencyRatio.setRatioType(RatioType.AGENCY);
                agencyRatio.setUser(null);
                agencyRatio.setPercentage(companyRatio);

                ratioRepository.save(agencyRatio);
            }
        }

        contractRepository.save(contract);
        return contract.getId();
    }

    public void changeStatus(Long id, ContractStatus next, ContractCategory category) {
        if (category == ContractCategory.INTERNAL) {
            InternalContract internal = internalContractRepository.findById(id)
                    .orElseThrow(() -> new EntityNotFoundException("내부 계약 없음"));

            if (!canTransition(internal.getStatus(), next)) {
                throw new IllegalStateException("잘못된 상태 변경");
            }

            internal.setStatus(next);
            internalContractRepository.save(internal);

        } else {
            Contract contract = contractRepository.findById(id)
                    .orElseThrow(() -> new EntityNotFoundException("외부 계약 없음"));

            if (!canTransition(contract.getStatus(), next)) {
                throw new IllegalStateException("잘못된 상태 변경");
            }

            contract.setStatus(next);
            contractRepository.save(contract);
        }
    }

//    public String signContract(Long contractId, SecurityUser user) throws IOException {
//        Contract contract = contractRepository.findById(contractId)
//                .orElseThrow(() -> new IllegalArgumentException("계약 없음"));
//        System.out.println("✔ 계약 조회 완료: " + contract.getTitle());
//
//        byte[] fileBytes = downloadFileFromUrl(contract.getFileUrl());
//        System.out.println("✔ 파일 다운로드 완료");
//
//        String fileName = extractFileNameFromUrl(contract.getFileUrl());
//        File tempFile = new File(System.getProperty("java.io.tmpdir"), fileName);
//        try (FileOutputStream fos = new FileOutputStream(tempFile)) {
//            fos.write(fileBytes);
//        }
//        System.out.println("✔ 임시 파일 생성 완료: " + tempFile.getAbsolutePath());
//
//        String signatureUrl = docusignService.sendEnvelope(tempFile, user.getNickname(), user.getEmail());
//        System.out.println("✔ DocuSign 서명 URL 발급 완료");
//
//        contract.setDocusignUrl(signatureUrl);
//        contract.setSignerName(user.getNickname());
//        contract.setStatus(ContractStatus.SIGNED);
//        contractRepository.save(contract);
//
//        return signatureUrl;
//    }

//    public String signContract(Long contractId, SecurityUser user) throws IOException {
//        InternalContract contract = internalContractRepository.findById(contractId)
//                .orElseThrow(() -> new IllegalArgumentException("계약 없음"));
//        System.out.println("✔ 계약 조회 완료: " + contract.getTitle());
//
//        byte[] fileBytes = downloadFileFromUrl(contract.getFileUrl());
//        System.out.println("✔ 파일 다운로드 완료");
//
//        String fileName = extractFileNameFromUrl(contract.getFileUrl());
//        File tempFile = new File(System.getProperty("java.io.tmpdir"), fileName);
//        try (FileOutputStream fos = new FileOutputStream(tempFile)) {
//            fos.write(fileBytes);
//        }
//        System.out.println("✔ 임시 파일 생성 완료: " + tempFile.getAbsolutePath());
//
//        String signatureUrl = docusignService.sendEnvelope(tempFile, user.getNickname(), user.getEmail());
//        System.out.println("✔ DocuSign 서명 URL 발급 완료");
//
//        contract.setDocusignUrl(signatureUrl);
//        contract.setSignerName(user.getNickname());
//        contract.setStatus(ContractStatus.SIGNED);
//        internalContractRepository.save(contract);
//
//        return signatureUrl;
//    }

    public String signContract(Long contractId, User signer) throws IOException {
        InternalContract contract = internalContractRepository.findById(contractId)
                .orElseThrow(() -> new IllegalArgumentException("계약 없음"));
        System.out.println("✔ 계약 조회 완료: " + contract.getTitle());

        byte[] fileBytes = downloadFileFromUrl(contract.getFileUrl());
        System.out.println("✔ 파일 다운로드 완료");

        String fileName = extractFileNameFromUrl(contract.getFileUrl());
        File tempFile = new File(System.getProperty("java.io.tmpdir"), fileName);
        try (FileOutputStream fos = new FileOutputStream(tempFile)) {
            fos.write(fileBytes);
        }
        System.out.println("✔ 임시 파일 생성 완료: " + tempFile.getAbsolutePath());

        String signatureUrl = docusignService.sendEnvelope(tempFile, signer.getNickname(), signer.getEmail());
        System.out.println("✔ DocuSign 서명 URL 발급 완료");

        contract.setDocusignUrl(signatureUrl);
        contract.setSignerName(signer.getNickname());
        contract.setStatus(ContractStatus.SIGNED);
        internalContractRepository.save(contract);

        return signatureUrl;
    }

    private byte[] downloadFileFromUrl(String fileUrl) {
        try {
            HttpClient client = HttpClient.newHttpClient();
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(fileUrl))
                    .GET()
                    .build();

            HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());

            if (response.statusCode() != 200) {
                throw new RuntimeException("파일 다운로드 실패. 상태 코드: " + response.statusCode());
            }

            return response.body();
        } catch (Exception e) {
            throw new RuntimeException("파일 다운로드 중 오류 발생", e);
        }
    }

    // url에서 파일 이름만 추출
    private String extractFileNameFromUrl(String url) {
        return url.substring(url.lastIndexOf("/") + 1);
    }

    public void signOfflineAsAdmin(Long id, String signerName, String memo) {
        Contract contract = contractRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("계약 없음"));

        if (contract.getContractType() != ContractType.PAPER) {
            throw new IllegalStateException("종이 계약이 아닙니다.");
        }

        if (contract.getStatus() != ContractStatus.REVIEW) {
            throw new IllegalStateException("검토 상태만 서명할 수 있습니다.");
        }

        contract.setSignerName(signerName);
        contract.setSignerMemo("오프라인 서명 메모: " + memo);
        contract.setStatus(ContractStatus.SIGNED_OFFLINE);

        contractRepository.save(contract);
    }


    private boolean canTransition(ContractStatus current, ContractStatus next) {
        return switch (current) {
            case DRAFT -> next == ContractStatus.REVIEW;
            case REVIEW -> next == ContractStatus.SIGNED || next == ContractStatus.SIGNED_OFFLINE;
            case SIGNED_OFFLINE, SIGNED -> next == ContractStatus.CONFIRMED;
            default -> false;
        };
    }

    public String getContractFileUrl(Long id, ContractCategory category) {
        switch (category) {
            case EXTERNAL -> {
                Contract contract = contractRepository.findById(id)
                        .orElseThrow(() -> new ApiException(ErrorCode.CONTRACT_NOT_FOUND));
                return contract.getFileUrl();
            }
            case INTERNAL -> {
                InternalContract internal = internalContractRepository.findById(id)
                        .orElseThrow(() -> new ApiException(ErrorCode.CONTRACT_NOT_FOUND));
                return internal.getFileUrl();
            }
            default -> throw new IllegalArgumentException("지원하지 않는 계약 카테고리입니다.");
        }
    }

    public Page<ContractResponse> getContractsByFilter(ContractSearchCondition condition, Pageable pageable) {
        return contractRepository.findAll(
                ContractSpecification.build(condition), pageable
        ).map(ContractResponse::from);
    }

    public ContractConditionResponse getContractConditions(Long id) {
        Contract contract = contractRepository.findById(id)
                .orElseThrow(() -> new ApiException(ErrorCode.CONTRACT_NOT_FOUND));
        return ContractConditionResponse.from(contract);
    }

    public List<ContractResponse> getExpiringContracts(ContractCategory category) {
        LocalDate deadline = LocalDate.now().plusDays(30);

        return switch (category) {
            case EXTERNAL -> {
                List<Contract> contracts = contractRepository.findExpiringContracts(deadline, ContractStatus.TERMINATED);
                yield contracts.stream()
                        .map(ContractResponse::from)
                        .toList();
            }
            case INTERNAL -> {
                List<InternalContract> internals = internalContractRepository.findExpiringContracts(deadline, ContractStatus.TERMINATED);
                yield internals.stream()
                        .map(ContractResponse::fromInternal)
                        .toList();
            }
            default -> throw new IllegalArgumentException("지원하지 않는 계약 카테고리입니다.");
        };
    }

    public Long getMonthlyContractCount() {
        LocalDate now = LocalDate.now();
        LocalDate startOfMonth = now.withDayOfMonth(1);
        LocalDate endOfMonth = now.withDayOfMonth(now.lengthOfMonth());

        return contractRepository.countByCreatedAtBetween(startOfMonth.atStartOfDay(), endOfMonth.atTime(LocalTime.MAX));
    }

    public Map<ContractStatus, Long> getContractStatusSummary() {
        List<Object[]> result = contractRepository.countContractsByStatus();
        return result.stream()
                .collect(Collectors.toMap(
                        row -> (ContractStatus) row[0],
                        row -> (Long) row[1]
                ));
    }

    @Transactional
    public void deleteContract(Long contractId, ContractCategory category) {

        switch (category) {
            case EXTERNAL -> {
                Contract contract = contractRepository.findById(contractId)
                        .orElseThrow(() -> new EntityNotFoundException("계약서를 찾을 수 없습니다."));
                // Soft delete: 상태만 변경
                contract.setStatus(ContractStatus.TERMINATED);

                // 필요 시: 연결된 SettlementRatio도 soft delete 처리
                List<SettlementRatio> ratios = ratioRepository.findByContract(contract);
                for (SettlementRatio ratio : ratios) {
                    ratio.setStatus(SettlementStatus.DELETED);
                }
            }

            case INTERNAL -> {
                InternalContract internalcontract = internalContractRepository.findById(contractId)
                        .orElseThrow(() -> new EntityNotFoundException("계약서를 찾을 수 없습니다."));
                // 내부 계약은 삭제
                internalcontract.setStatus(ContractStatus.TERMINATED);
            }
        }
    }
}

File: fe/src/features/finance-legal/services/contractService.ts
import {
  ContractCategory,
  ContractStatus,
  CreateContractRequest,
  CreateInternalContractRequest,
  UpdateContractRequest,
  ChangeStatusRequest,
  OfflineSignRequest,
  ContractSimpleDto,
  ContractDetailDto,
  ContractResponse,
} from "../types/Contract";

const API_BASE_URL = "http://localhost:8080/api/v1/legal";

// 계약서 생성 (외부 계약)
export const createContract = async (
  request: CreateContractRequest,
  file: File
): Promise<number> => {
  const formData = new FormData();
  formData.append(
    "request",
    new Blob([JSON.stringify(request)], { type: "application/json" })
  );
  formData.append("file", file);

  const response = await fetch(`${API_BASE_URL}/contracts`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    throw new Error("계약서 생성에 실패했습니다.");
  }

  return await response.json();
};

// 내부 계약 생성
export const createInternalContract = async (
  request: CreateInternalContractRequest,
  file: File
): Promise<number> => {
  const formData = new FormData();
  formData.append(
    "request",
    new Blob([JSON.stringify(request)], { type: "application/json" })
  );
  formData.append("file", file);

  const response = await fetch(`${API_BASE_URL}/internal-contracts`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    throw new Error("내부 계약 생성에 실패했습니다.");
  }

  return await response.json();
};

// 계약서 상태 변경
export const changeContractStatus = async (
  id: number,
  request: ChangeStatusRequest,
  category: ContractCategory
): Promise<void> => {
  const response = await fetch(
    `${API_BASE_URL}/${id}/status?category=${category}`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    }
  );

  if (!response.ok) {
    throw new Error("계약서 상태 변경에 실패했습니다.");
  }
};

// 오프라인 서명 처리 (외부 계약자용)
export const signOfflineAsAdmin = async (
  id: number,
  request: OfflineSignRequest
): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/${id}/sign-offline`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error("오프라인 서명 처리에 실패했습니다.");
  }
};

// 계약서 전자 서명 요청 생성 (대리)
export const signOnBehalf = async (
  id: number,
  userId: number
): Promise<string> => {
  const response = await fetch(`${API_BASE_URL}/${id}/sign?userId=${userId}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("전자 서명 요청 생성에 실패했습니다.");
  }

  return await response.text();
};

// 특정 유저 계약서 리스트 조회
export const getContractsByUser = async (
  userId: number,
  category: ContractCategory,
  page: number = 0,
  size: number = 10
): Promise<ContractSimpleDto[]> => {
  const params = new URLSearchParams({
    userId: userId.toString(),
    category: category,
    page: page.toString(),
    size: size.toString(),
  });

  const response = await fetch(`${API_BASE_URL}/by-user?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("사용자 계약서 목록 조회에 실패했습니다.");
  }

  return await response.json();
};

// 계약 상세 조회
export const getContractDetail = async (
  id: number,
  category: ContractCategory
): Promise<ContractDetailDto> => {
  const params = new URLSearchParams({
    category: category,
  });

  const response = await fetch(`${API_BASE_URL}/${id}?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("계약서 상세 조회에 실패했습니다.");
  }

  return await response.json();
};

// 모든 계약서 리스트 조회 (페이징 포함)
export const getAllContracts = async (
  category: ContractCategory,
  page: number = 0,
  size: number = 10
): Promise<{
  content: ContractSimpleDto[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}> => {
  const params = new URLSearchParams({
    category: category,
    page: page.toString(),
    size: size.toString(),
  });

  const response = await fetch(`${API_BASE_URL}?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("계약서 목록 조회에 실패했습니다.");
  }

  return await response.json();
};

// 계약서 최종 확정
export const confirmContract = async (
  id: number,
  category: ContractCategory
): Promise<void> => {
  const params = new URLSearchParams({
    category: category,
  });

  const response = await fetch(`${API_BASE_URL}/${id}/confirm?${params}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("계약서 확정에 실패했습니다.");
  }
};

// 계약서 파일 URL 조회
export const getContractFileUrl = async (
  id: number,
  category: ContractCategory
): Promise<string> => {
  const params = new URLSearchParams({
    category: category,
  });

  const response = await fetch(`${API_BASE_URL}/${id}/file-url?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("계약서 파일 URL 조회에 실패했습니다.");
  }

  return await response.text();
};

// 만료 예정 계약 조회
export const getExpiringContracts = async (
  category: ContractCategory
): Promise<ContractResponse[]> => {
  const params = new URLSearchParams({
    category: category,
  });

  const response = await fetch(`${API_BASE_URL}/expiring?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("만료 예정 계약 조회에 실패했습니다.");
  }

  return await response.json();
};

// 계약서 수정
export const updateContract = async (
  contractId: number,
  request: UpdateContractRequest,
  file?: File
): Promise<number> => {
  const formData = new FormData();
  formData.append(
    "request",
    new Blob([JSON.stringify(request)], { type: "application/json" })
  );

  if (file) {
    formData.append("file", file);
  }

  const response = await fetch(`${API_BASE_URL}/${contractId}`, {
    method: "PUT",
    body: formData,
  });

  if (!response.ok) {
    throw new Error("계약서 수정에 실패했습니다.");
  }

  return await response.json();
};

// 계약서 삭제
export const deleteContract = async (
  contractId: number,
  category: ContractCategory
): Promise<void> => {
  const params = new URLSearchParams({
    category: category,
  });

  const response = await fetch(`${API_BASE_URL}/${contractId}?${params}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error("계약서 삭제에 실패했습니다.");
  }
};

// 계약서 서비스 객체
export const contractService = {
  createContract,
  createInternalContract,
  changeContractStatus,
  signOfflineAsAdmin,
  signOnBehalf,
  getContractsByUser,
  getContractDetail,
  getAllContracts,
  confirmContract,
  getContractFileUrl,
  getExpiringContracts,
  updateContract,
  deleteContract,
};
